README.md
# Park254 Parking App Server
[![Node.js CI](https://github.com/Developer-Square/Park254_Backend/actions/workflows/node.js.yml/badge.svg)](https://github.com/Developer-Square/Park254_Backend/actions/workflows/node.js.yml)

## Manual Installation

Clone the repo:

```bash
git clone --depth 1 https://github.com/Developer-Square/Park254_Backend
cd Park254_Backend
```

Install the dependencies:

```bash
yarn install
```

Set the environment variables:

```bash
cp .env.example .env

# open .env and modify the environment variables (if needed)
```

## Table of Contents

- [Features](#features)
- [Commands](#commands)
- [Environment Variables](#environment-variables)
- [Project Structure](#project-structure)
- [API Documentation](#api-documentation)
- [Error Handling](#error-handling)
- [Validation](#validation)
- [Authentication](#authentication)
- [Authorization](#authorization)
- [Logging](#logging)
- [Custom Mongoose Plugins](#custom-mongoose-plugins)
- [Linting](#linting)

## Features

- **ES9**: latest ECMAScript features
- **NoSQL database**: [MongoDB](https://www.mongodb.com) object data modeling using [Mongoose](https://mongoosejs.com)
- **Authentication and authorization**: using [passport](http://www.passportjs.org)
- **Validation**: request data validation using [Joi](https://github.com/hapijs/joi)
- **Logging**: using [winston](https://github.com/winstonjs/winston) and [morgan](https://github.com/expressjs/morgan)
- **Testing**: unit and integration tests using [Jest](https://jestjs.io)
- **Error handling**: centralized error handling mechanism
- **API documentation**: with [swagger-jsdoc](https://github.com/Surnet/swagger-jsdoc) and [swagger-ui-express](https://github.com/scottie1984/swagger-ui-express)
- **Process management**: advanced production process management using [PM2](https://pm2.keymetrics.io)
- **Dependency management**: with [Yarn](https://yarnpkg.com)
- **Environment variables**: using [dotenv](https://github.com/motdotla/dotenv) and [cross-env](https://github.com/kentcdodds/cross-env#readme)
- **Security**: set security HTTP headers using [helmet](https://helmetjs.github.io)
- **Santizing**: sanitize request data against xss and query injection
- **CORS**: Cross-Origin Resource-Sharing enabled using [cors](https://github.com/expressjs/cors)
- **Compression**: gzip compression with [compression](https://github.com/expressjs/compression)
- **CI**: continuous integration with [Travis CI](https://travis-ci.org)
- **Docker support**
- **Code coverage**: using [coveralls](https://coveralls.io)
- **Code quality**: with [Codacy](https://www.codacy.com)
- **Git hooks**: with [husky](https://github.com/typicode/husky) and [lint-staged](https://github.com/okonet/lint-staged)
- **Linting**: with [ESLint](https://eslint.org) and [Prettier](https://prettier.io)
- **Editor config**: consistent editor configuration using [EditorConfig](https://editorconfig.org)

## Commands

Running locally:

```bash
yarn dev
```

Running in production:

```bash
yarn start
```

Testing:

```bash
# run all tests
yarn test

# run all tests in watch mode
yarn test:watch

# run test coverage
yarn coverage
```

Docker:

```bash
# run docker container in development mode
yarn docker:dev

# run docker container in production mode
yarn docker:prod

# run all tests in a docker container
yarn docker:test
```

Linting:

```bash
# run ESLint
yarn lint

# fix ESLint errors
yarn lint:fix

# run prettier
yarn prettier

# fix prettier errors
yarn prettier:fix
```

## Environment Variables

The environment variables can be found and modified in the `.env` file. They come with these default values:

```bash
# Port number
PORT=3000

# URL of the Mongo DB
MONGODB_URL=mongodb://127.0.0.1:27017/Park254_Backend

# JWT
# JWT secret key
JWT_SECRET=thisisasamplesecret
# Number of minutes after which an access token expires
JWT_ACCESS_EXPIRATION_MINUTES=30
# Number of days after which a refresh token expires
JWT_REFRESH_EXPIRATION_DAYS=30

# SMTP configuration options for the email service
# For testing, you can use a fake SMTP service like Ethereal: https://ethereal.email/create
SMTP_HOST=email-server
SMTP_PORT=587
SMTP_USERNAME=email-server-username
SMTP_PASSWORD=email-server-password
EMAIL_FROM=support@yourapp.com
```

## Project Structure

```
src\
 |--config\         # Environment variables and configuration related things
 |--controllers\    # Route controllers (controller layer)
 |--docs\           # Swagger files
 |--middlewares\    # Custom express middlewares
 |--models\         # Mongoose models (data layer)
 |--routes\         # Routes
 |--services\       # Business logic (service layer)
 |--utils\          # Utility classes and functions
 |--validations\    # Request data validation schemas
 |--app.js          # Express app
 |--index.js        # App entry point
```

## API Documentation

To view the list of available APIs and their specifications, run the server and go to `http://localhost:3000/v1/docs` in your browser. This documentation page is automatically generated using the [swagger](https://swagger.io/) definitions written as comments in the route files.

### API Endpoints

List of available routes:

**Auth routes**:\
`POST /v1/auth/register` - register\
`POST /v1/auth/login` - login\
`POST /v1/auth/refresh-tokens` - refresh auth tokens\
`POST /v1/auth/forgot-password` - send reset password email\
`POST /v1/auth/reset-password` - reset password

**User routes**:\
`POST /v1/users` - create a user\
`GET /v1/users` - get all users\
`GET /v1/users/:userId` - get user\
`PATCH /v1/users/:userId` - update user\
`DELETE /v1/users/:userId` - delete user

**Vehicle routes**:\
`POST /v1/vehicles` - create a vehicle\
`GET /v1/vehicles` - get all vehicles\
`GET /v1/vehicles/:vehicleId` - get vehicle\
`PATCH /v1/vehicles/:vehicleId` - update vehicle\
`DELETE /v1/vehicles/:vehicleId` - delete vehicle

**Parking lot routes**:\
`POST /v1/parkingLots` - create a parking lot\
`GET /v1/nearbyParking` - get nearby parking lots\
`GET /v1/parkingLots` - get all parking lots\
`GET /v1/parkingLots/:parkingLotId` - get parking lot\
`PATCH /v1/parkingLots/:parkingLotId` - update parking lot\
`DELETE /v1/parkingLots/:parkingLotId` - delete parking lot

**Rating routes**:\
`POST /v1/ratings` - create a rating\
`GET /v1/ratings` - get all ratings\
`GET /v1/ratings/:ratingId` - get rating\
`DELETE /v1/ratings/:ratingId` - delete rating

**Booking routes**:\
`POST /v1/bookings` - book a parking lot\
`GET /v1/bookings` - get all bookings\
`GET /v1/bookings/:userId` - get booking by id\
`PATCH /v1/bookings/:userId` - update booking by id\
`DELETE /v1/bookings/:userId` - delete booking by id\
`POST /v1/bookings/:userId` - cancel booking by id\
`POST /v1/spaces` - find available spaces

**MPESA routes**:\
`POST /v1/mpesa` - make MPESA payment\
`GET /v1/mpesa` - get all transactions\
`GET /v1/mpesa/:transactionId` - get transaction by id\
`DELETE /v1/mpesa/:transactionId` - delete transaction by id\
`POST /v1/paymentCallback` - callback url for mpesa transactions\
`GET /v1/paymentCallback` - fetch transaction using phoneNumber, amount, and createdAt

## Error Handling

The app has a centralized error handling mechanism.

Controllers should try to catch the errors and forward them to the error handling middleware (by calling `next(error)`). For convenience, you can also wrap the controller inside the catchAsync utility wrapper, which forwards the error.

```javascript
const catchAsync = require('../utils/catchAsync');

const controller = catchAsync(async (req, res) => {
  // this error will be forwarded to the error handling middleware
  throw new Error('Something wrong happened');
});
```

The error handling middleware sends an error response, which has the following format:

```json
{
  "code": 404,
  "message": "Not found"
}
```

When running in development mode, the error response also contains the error stack.

The app has a utility ApiError class to which you can attach a response code and a message, and then throw it from anywhere (catchAsync will catch it).

For example, if you are trying to get a user from the DB who is not found, and you want to send a 404 error, the code should look something like:

```javascript
const httpStatus = require('http-status');
const ApiError = require('../utils/ApiError');
const User = require('../models/User');

const getUser = async (userId) => {
  const user = await User.findById(userId);
  if (!user) {
    throw new ApiError(httpStatus.NOT_FOUND, 'User not found');
  }
};
```

## Validation

Request data is validated using [Joi](https://joi.dev/). Check the [documentation](https://joi.dev/api/) for more details on how to write Joi validation schemas.

The validation schemas are defined in the `src/validations` directory and are used in the routes by providing them as parameters to the `validate` middleware.

```javascript
const express = require('express');
const validate = require('../../middlewares/validate');
const userValidation = require('../../validations/user.validation');
const userController = require('../../controllers/user.controller');

const router = express.Router();

router.post('/users', validate(userValidation.createUser), userController.createUser);
```

## Authentication

To require authentication for certain routes, you can use the `auth` middleware.

```javascript
const express = require('express');
const auth = require('../../middlewares/auth');
const userController = require('../../controllers/user.controller');

const router = express.Router();

router.post('/users', auth(), userController.createUser);
```

These routes require a valid JWT access token in the Authorization request header using the Bearer schema. If the request does not contain a valid access token, an Unauthorized (401) error is thrown.

**Generating Access Tokens**:

An access token can be generated by making a successful call to the register (`POST /v1/auth/register`) or login (`POST /v1/auth/login`) endpoints. The response of these endpoints also contains refresh tokens (explained below).

An access token is valid for 30 minutes. You can modify this expiration time by changing the `JWT_ACCESS_EXPIRATION_MINUTES` environment variable in the .env file.

**Refreshing Access Tokens**:

After the access token expires, a new access token can be generated, by making a call to the refresh token endpoint (`POST /v1/auth/refresh-tokens`) and sending along a valid refresh token in the request body. This call returns a new access token and a new refresh token.

A refresh token is valid for 30 days. You can modify this expiration time by changing the `JWT_REFRESH_EXPIRATION_DAYS` environment variable in the .env file.

## Authorization

The `auth` middleware can also be used to require certain rights/permissions to access a route.

```javascript
const express = require('express');
const auth = require('../../middlewares/auth');
const userController = require('../../controllers/user.controller');

const router = express.Router();

router.post('/users', auth('manageUsers'), userController.createUser);
```

In the example above, an authenticated user can access this route only if that user has the `manageUsers` permission.

The permissions are role-based. You can view the permissions/rights of each role in the `src/config/roles.js` file.

If the user making the request does not have the required permissions to access this route, a Forbidden (403) error is thrown.

## Logging

Import the logger from `src/config/logger.js`. It is using the [Winston](https://github.com/winstonjs/winston) logging library.

Logging should be done according to the following severity levels (ascending order from most important to least important):

```javascript
const logger = require('<path to src>/config/logger');

logger.error('message'); // level 0
logger.warn('message'); // level 1
logger.info('message'); // level 2
logger.http('message'); // level 3
logger.verbose('message'); // level 4
logger.debug('message'); // level 5
```

In development mode, log messages of all severity levels will be printed to the console.

In production mode, only `info`, `warn`, and `error` logs will be printed to the console.\
It is up to the server (or process manager) to actually read them from the console and store them in log files.\
This app uses pm2 in production mode, which is already configured to store the logs in log files.

Note: API request information (request url, response code, timestamp, etc.) are also automatically logged (using [morgan](https://github.com/expressjs/morgan)).

## Custom Mongoose Plugins

The app also contains 2 custom mongoose plugins that you can attach to any mongoose model schema. You can find the plugins in `src/models/plugins`.

```javascript
const mongoose = require('mongoose');
const { toJSON, paginate } = require('./plugins');

const userSchema = mongoose.Schema(
  {
    /* schema definition here */
  },
  { timestamps: true }
);

userSchema.plugin(toJSON);
userSchema.plugin(paginate);

const User = mongoose.model('User', userSchema);
```

### toJSON

The toJSON plugin applies the following changes in the toJSON transform call:

- removes \_\_v, createdAt, updatedAt, and any schema path that has private: true
- replaces \_id with id

### paginate

The paginate plugin adds the `paginate` static method to the mongoose schema.

Adding this plugin to the `User` model schema will allow you to do the following:

```javascript
const queryUsers = async (filter, options) => {
  const users = await User.paginate(filter, options);
  return users;
};
```

The `filter` param is a regular mongo filter.

The `options` param can have the following (optional) fields:

```javascript
const options = {
  sortBy: 'name:desc', // sort order
  limit: 5, // maximum results per page
  page: 2, // page number
};
```

The plugin also supports sorting by multiple criteria (separated by a comma): `sortBy: name:desc,role:asc`

The `paginate` method returns a Promise, which fulfills with an object having the following properties:

```json
{
  "results": [],
  "page": 2,
  "limit": 5,
  "totalPages": 10,
  "totalResults": 48
}
```

## Linting

Linting is done using [ESLint](https://eslint.org/) and [Prettier](https://prettier.io).

In this app, ESLint is configured to follow the [Airbnb JavaScript style guide](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb-base) with some modifications. It also extends [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) to turn off all rules that are unnecessary or might conflict with Prettier.

To modify the ESLint configuration, update the `.eslintrc.json` file. To modify the Prettier configuration, update the `.prettierrc.json` file.

To prevent a certain file or directory from being linted, add it to `.eslintignore` and `.prettierignore`.

To maintain a consistent coding style across different IDEs, the project contains `.editorconfig`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.4](https://github.com/saisilinus/Park254_Backend/compare/v0.0.3...v0.0.4) (2022-11-26)


### Bug Fixes

* remove .env ([ac2c9b3](https://github.com/saisilinus/Park254_Backend/commit/ac2c9b352904149a52f0ca5c6d357a1dc893cd09))

### [0.0.3](https://github.com/saisilinus/Park254_Backend/compare/v0.0.2...v0.0.3) (2022-11-26)


### Features

* add auth + validation for spaces ([c413b14](https://github.com/saisilinus/Park254_Backend/commit/c413b14716bed1385842f4e897980ff178fd7b6a))
* add booking ([3dd0834](https://github.com/saisilinus/Park254_Backend/commit/3dd083438de4750ddd6d1b58549385bec166e26a))
* add booking population ([053652d](https://github.com/saisilinus/Park254_Backend/commit/053652dbda250721c716414a0351b8b63b595102))
* add booking tests ([da6c5de](https://github.com/saisilinus/Park254_Backend/commit/da6c5de7296cc5b127799d710ee4ba44e24ceedb))
* add booking tests ([b3ebb93](https://github.com/saisilinus/Park254_Backend/commit/b3ebb935f417de50160dca7332c621c144e6dfe2))
* add confirmParkingSpaces ([4587f37](https://github.com/saisilinus/Park254_Backend/commit/4587f37839d01df0e9f9fc0c27e931be252ef05e))
* add default values for Features ([0110621](https://github.com/saisilinus/Park254_Backend/commit/0110621b3b2a35876561ac98718d71f34b81d986))
* add fetch transaction ([89e517a](https://github.com/saisilinus/Park254_Backend/commit/89e517aa63b67a2274eb09424511baf8760b599e))
* add find available spaces ([c2e76fc](https://github.com/saisilinus/Park254_Backend/commit/c2e76fc6ebc0af522987d77fb040a6ec8695c211))
* add GitHub CI badge ([78c7608](https://github.com/saisilinus/Park254_Backend/commit/78c7608d14fec5893f1d148317bf49c23c51a0b7))
* add landing page ([0bb787d](https://github.com/saisilinus/Park254_Backend/commit/0bb787d6f403f5a5b7b9f5450465be9fd3fd29b9))
* add mongoclient and remove mongoose ([b98c086](https://github.com/saisilinus/Park254_Backend/commit/b98c086df9653608baba1f11db281fea984d5d56))
* add MPESA production config ([5925381](https://github.com/saisilinus/Park254_Backend/commit/5925381dd76c901dda9edb33626081a9069e4822))
* add mPesa utility functions ([4eef07d](https://github.com/saisilinus/Park254_Backend/commit/4eef07de132d78b6727073928f8a2e0a89f207ce))
* add mpesa webhook ([b27d57c](https://github.com/saisilinus/Park254_Backend/commit/b27d57ce7c758cdf5c0da99ed23fde7e4053177f))
* add parking lot features ([d8d7d54](https://github.com/saisilinus/Park254_Backend/commit/d8d7d54f88542e26a4c2d3d2033716529fd24f30))
* add phone + vehicles to users ([b373005](https://github.com/saisilinus/Park254_Backend/commit/b373005c2c4627b033437c51c117b8f6f2474a08))
* add phone login ([b5e32f3](https://github.com/saisilinus/Park254_Backend/commit/b5e32f333069a6b1c8a9b07c3dccf99d8889d12f))
* add population for owner in parkinglot ([f19bdf1](https://github.com/saisilinus/Park254_Backend/commit/f19bdf162bcd1ec15c1282c1ab99f24c2785c210))
* add procfile ([86329e3](https://github.com/saisilinus/Park254_Backend/commit/86329e35866561ab1b5ba59fc423a5cbc4a8359d))
* add production config ([8e2d755](https://github.com/saisilinus/Park254_Backend/commit/8e2d755924f623d33f0c546302c7018aafee2a22))
* add refresh-site api ([1814628](https://github.com/saisilinus/Park254_Backend/commit/18146282a325e26b3e8bbd414697ae7678ec6b62))
* add tests for phone + vehicles ([6ac9bf7](https://github.com/saisilinus/Park254_Backend/commit/6ac9bf7f3939720367e4d13922b847ef960f3c71))
* add transaction tests ([213f320](https://github.com/saisilinus/Park254_Backend/commit/213f3204e90c1bed3735395f0f60d7a0a8db1750))
* add validation + auth to booking routes ([e3ad31a](https://github.com/saisilinus/Park254_Backend/commit/e3ad31a83f7b10284df4f6e82587a2a71bd51864))
* add validation + auth to mpesa ([0c7bede](https://github.com/saisilinus/Park254_Backend/commit/0c7bede24731644b823b4567e529f92e65c595b9))
* add vehicle + phone fields ([79f4f08](https://github.com/saisilinus/Park254_Backend/commit/79f4f082c29bb26e4d0b53c7036d54e7ec80b2d2))
* add vehicle tests ([5e7a7f3](https://github.com/saisilinus/Park254_Backend/commit/5e7a7f377e9defa309b9df472f9ec9baa37b699a))
* add verifyUser + booking tests ([38e9a65](https://github.com/saisilinus/Park254_Backend/commit/38e9a65e176179c68b3cf9283deb850ca2deecd0))
* add verifyUser tests for ratings + lots ([711d47b](https://github.com/saisilinus/Park254_Backend/commit/711d47b859811861ca0d0ed577de55b052f668af))
* change node version to 18 ([60c1e89](https://github.com/saisilinus/Park254_Backend/commit/60c1e897348df52a8f379fb03caaf959ccd89b39))
* change production mongodb url ([dc22a5c](https://github.com/saisilinus/Park254_Backend/commit/dc22a5cf95632d2cb09068cebd776f5b0580e2e9))
* change production mongodb url ([237417e](https://github.com/saisilinus/Park254_Backend/commit/237417e79e37942749ab39b169f4a7e8b06b9016))
* changed the start script in Procfile ([8eed47a](https://github.com/saisilinus/Park254_Backend/commit/8eed47ac4fdd5aa4ae528de5411a2df71050fe94))
* merge parking lot fields ([f32477d](https://github.com/saisilinus/Park254_Backend/commit/f32477d709518c9ab3c3305d0c34ce1620e4be3b))
* mpesa integration ([1fbcd6e](https://github.com/saisilinus/Park254_Backend/commit/1fbcd6e84c881a715f2951fdd670a6a61308ae61))
* remove Travis CI ([9a69aad](https://github.com/saisilinus/Park254_Backend/commit/9a69aadb79e15bd8fcbe4b224bd71f71e4356aa6))
* smtp config ([1ec45d5](https://github.com/saisilinus/Park254_Backend/commit/1ec45d56b16feb6e3b0bff5780fddd51cd5d9601))


### Bug Fixes

* add console.log for production url ([1e764fb](https://github.com/saisilinus/Park254_Backend/commit/1e764fbd62a06c0dc8524797ac8ec0007c23670c))
* add dotenv to index.js ([be9c9ec](https://github.com/saisilinus/Park254_Backend/commit/be9c9ec03450d4be9f2fd0c5351038738c9b1d9e))

src/config/config.js
/* eslint-disable no-console */
const dotenv = require('dotenv');
const path = require('path');
const Joi = require('joi');
const production = require('./production');

dotenv.config({ path: path.join(__dirname, '../../.env') });

const envVarsSchema = Joi.object()
  .keys({
    NODE_ENV: Joi.string().valid('production', 'development', 'test').required(),
    PORT: Joi.number().default(3000),
    MONGODB_URL: Joi.string().required().description('Mongo DB url'),
    JWT_SECRET: Joi.string().required().description('JWT secret key'),
    JWT_ACCESS_EXPIRATION_MINUTES: Joi.number().default(30).description('minutes after which access tokens expire'),
    JWT_REFRESH_EXPIRATION_DAYS: Joi.number().default(30).description('days after which refresh tokens expire'),
    SMTP_HOST: Joi.string().description('server that will send the emails'),
    SMTP_PORT: Joi.number().description('port to connect to the email server'),
    SMTP_USERNAME: Joi.string().description('username for email server'),
    SMTP_PASSWORD: Joi.string().description('password for email server'),
    EMAIL_FROM: Joi.string().description('the from field in the emails sent by the app'),
    CONSUMER_KEY: Joi.string().required().description('Mpesa consumer key'),
    CONSUMER_SECRET: Joi.string().required().description('Mpesa consumer secret'),
    MPESA_PASSKEY_SANDBOX: Joi.string().required().description('Mpesa passkey'),
  })
  .unknown();

const { value: envVars, error } = envVarsSchema.prefs({ errors: { label: 'key' } }).validate(process.env);

if (error) {
  throw new Error(`Config validation error: ${error.message}`);
}

module.exports = {
  env: envVars.NODE_ENV,
  port: envVars.PORT,
  mongoose: {
    url:
      envVars.NODE_ENV === 'production'
        ? production.MONGODB_ATLAS_URL
        : envVars.MONGODB_URL + (envVars.NODE_ENV === 'test' ? '-test' : ''),
    options: {
      useCreateIndex: true,
      useNewUrlParser: true,
      useUnifiedTopology: true,
    },
  },
  jwt: {
    secret: envVars.NODE_ENV === 'production' ? production.JWT_SECRET : envVars.JWT_SECRET,
    accessExpirationMinutes: envVars.JWT_ACCESS_EXPIRATION_MINUTES,
    refreshExpirationDays: envVars.JWT_REFRESH_EXPIRATION_DAYS,
    resetPasswordExpirationMinutes: 10,
  },
  email: {
    smtp: {
      host: envVars.SMTP_HOST,
      port: envVars.SMTP_PORT,
      auth: {
        user: envVars.NODE_ENV === 'production' ? production.SMTP_USERNAME : envVars.SMTP_USERNAME,
        pass: envVars.NODE_ENV === 'production' ? production.SMTP_PASSWORD : envVars.SMTP_PASSWORD,
      },
    },
    from: envVars.EMAIL_FROM,
  },
  mpesa: {
    consumer_key: envVars.NODE_ENV === 'production' ? production.MPESA_CONSUMER_KEY : envVars.CONSUMER_KEY,
    consumer_secret: envVars.NODE_ENV === 'production' ? production.MPESA_CONSUMER_SECRET : envVars.CONSUMER_SECRET,
    passkey: envVars.NODE_ENV === 'production' ? production.MPESA_PASSKEY : envVars.CONSUMER_SECRET,
  },
};
